src/web/components/Onwards/Onwards.tsx

```diff
@@ -1,28 +1,170 @@
+    // a single slash - prevents malformed urls
+    const trimmed = parts
+        .map(part => {
+            // Trim left
+            if (part.substr(0, 1) === '/') return part.slice(1);
+            return part;
+        })
+        .map(part => {
+            // Trim right
+            if (part.substr(part.length - 1, 1) === '/')
+                return part.slice(0, -1);
+            return part;
+        });
+
+    return trimmed.join('/');
+};
+
+const firstPopularTag = (pageTags: string[], isPaidContent: boolean) => {
+    // This function looks for the first tag in pageTags, that also exists in our whitelist
+    if (!pageTags) {
+        // If there are no page tags we will never find a match so
+        return false;
+    }
+
+    const isInWhitelist = (tag: string) => WHITELISTED_TAGS.includes(tag);
+
+    // For paid content we just return the first tag, otherwise we
+    // filter for the first tag in the whitelist
+    return isPaidContent ? pageTags[0] : pageTags.find(isInWhitelist);
+};
+
 type Props = {
     ajaxUrl: string;
+    hasRelated: boolean;
+    hasStoryPackage: boolean;
+    isAdFreeUser: boolean;
     pageId: string;
-    pathId: OnwardsIdType;
-    component: React.ElementType;
+    isPaidContent: boolean;
+    showRelatedContent: boolean;
+    keywordIds: string[];
+    contentType: string;
 };
 
-export const Onwards = ({ ajaxUrl, pageId, pathId, component }: Props) => {
-    const url = `${new URL(`${pathId}/${pageId}`, ajaxUrl)}.json`;
-    const { data, error } = useApi<OnwardsType>(url);
+export const Onwards = ({
+    ajaxUrl,
+    hasRelated,
+    hasStoryPackage,
+    isAdFreeUser,
+    pageId,
+    isPaidContent,
+    showRelatedContent,
+    keywordIds,
+    contentType,
+}: Props) => {
+    const onwardSections: OnwardsType[] = [];
 
-    if (error) {
-        window.guardian.modules.sentry.reportError(error, 'story-package');
-        return null;
+    if (hasStoryPackage) {
+        // Always fetch the story package if it exists
+        const { data } = useApi(
+            joinUrl([ajaxUrl, 'story-package', `${pageId}.json?dcr=true`]),
+        );
+
+        const storyPackage = data;
+
+        if (data && data.trails && data.trails.length > 7) {
+            onwardSections.push({
+                heading: storyPackage.heading,
+                trails: storyPackage.trails,
+                layout: 'fourAndFour',
+            });
+        }
+    } else if (isAdFreeUser && isPaidContent) {
+        // Don't show any related content (other than story packages) for
+        // adfree users when the content is paid for
+    } else if (showRelatedContent && hasRelated) {
+        // Related content can be a collection of articles based on
+        // two things, 1: A popular tag, or 2: A generic text match
+        const popularTag = firstPopularTag(keywordIds, isPaidContent);
+
+        let relatedUrl;
+        if (popularTag) {
```

Comment:
It might be worth breaking out the url builder functions here into their own methods and making  this more functional `(path, isPopTag) => path`, `(path, isPaidContent, contentType) => path` etc. Making it more testable?

Reply:
I've refactored the logic to have one branch per possible outcome (removing the nested if statements) - is this closer to what you were thinking or?

Reply:
I find the more functional approach easier to reason with - named function, you know what params it takes and what is expected as output. You can test separately (but maybe you're testing too deep there). And you can compose those together.

Something like https://gist.github.com/JamieMason/172460a36a0eaef24233e6edb2706f83

```
const buildPath = compose(
	addPopularPath(path, isPopTag), // Returns a function that returns
	addPathExcludes(path, isPaidContent, contentType)
)

buildPath(path, isPopTag, IsPaidContent, contentType)
```

This is a minor, and maybe falls foul of simplest approach though... maybe the if statements are the more reasonable-with approach.

Reply:
Yeah, this code is extremely likely to be moved/changed in the near future so I'm keen to keep things as simple as possible. It's also nice (at this stage) to not stray too far from frontend - at least while we're in the flow of porting that logic over